### PR TITLE
Add CertLogic to evaluate national rules

### PR DESCRIPTION
**Pull request by ubamrein**
We add CertLogic to evaluate national rules. This is a _draft_ since we still use a custom branch for the CertLogic with various fixes on. 

As soon as the PRs on the CertLogic repos are merged we can push this further.

### CODE REVIEW

File: Sources/CovidCertificateSDK/CertLogic/CertLogic.swift [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    private static let vaccineImmunityKey: String = "vaccine-immunity"
+    private static let singleVaccineValidityOffsetKey: String = "single-vaccine-validity-offset"
+    private static let pcrTestValidityKey: String = "pcr-test-validity"
+    private static let ratTestValidityKey: String = "ratTestValidity"
+
+    var rules: [JSON] = []
+    var valueSets: JSON = []
+    let calendar: Calendar
+
+    public var maxValidity: Int64 { valueSets[CertLogic.acceptanceCriteriaKey][CertLogic.vaccineImmunityKey].int ?? 0 }
```
Comment (ubamrein) [inline Sources/CovidCertificateSDK/CertLogic/CertLogic.swift:37]: Needs to be nullable. Don't show invalid dates, but no date

File: Sources/CovidCertificateSDK/CertLogic/CertLogic.swift [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+    public func checkRules(hcert: EuHealthCert, validationClock: Date = Date()) -> Result<Void, CertLogicValidationError> {
+        var external = JSON(
+            ["validationClock": ISO8601DateFormatter().string(from: validationClock),
+             "validationClockAtStartOfDay": ISO8601DateFormatter().string(from: calendar.startOfDay(for: validationClock))]
+        )
+        external["valueSets"] = valueSets
+        var failedTests: [String: String] = [:]
+        guard let dgcJson = try? JSONEncoder().encode(hcert) else {
```
Comment (ubamrein) [inline Sources/CovidCertificateSDK/CertLogic/CertLogic.swift:67]: rename to dcc

File: Sources/CovidCertificateSDK/ChCovidCert.swift [hunk 3]
```diff
@@ -154,14 +155,81 @@ public struct ChCovidCert {
...
-    public func checkNationalRules(dgc: EuHealthCert, forceUpdate _: Bool, _ completionHandler: @escaping (Result<VerificationResult, NationalRulesError>) -> Void) {
+    public func checkNationalRules(dgc: EuHealthCert, forceUpdate: Bool, _ completionHandler: @escaping (Result<VerificationResult, NationalRulesError>) -> Void) {
         switch dgc.certType {
-        case .vaccination:
-            nationalRules.verifyVaccine(vaccine: dgc.vaccinations![0], completionHandler)
-        case .recovery:
-            nationalRules.verifyRecovery(recovery: dgc.pastInfections![0], completionHandler)
-        case .test:
-            nationalRules.verifyTest(test: dgc.tests![0], completionHandler)
+        case .vaccination, .recovery, .test:
```
Comment (ubamrein) [inline Sources/CovidCertificateSDK/ChCovidCert.swift:160]: maybe simplify

File: Sources/CovidCertificateSDK/ChCovidCert.swift [hunk 4]
```diff
@@ -154,14 +155,81 @@ public struct ChCovidCert {
...
+            trustListManager.nationalRulesListUpdater.addCheckOperation(forceUpdate: forceUpdate, checkOperation: { error in
+                if let e = error?.asNationalRulesError() {
+                    completionHandler(.failure(e))
+                    return
+                } else {
+                    let list = self.trustListManager.trustStorage.nationalRules()
+
+                    guard let result = CertLogic(),
```
Comment (ubamrein) [inline Sources/CovidCertificateSDK/ChCovidCert.swift:168]:
```suggestion
                    guard let certLogic = CertLogic(),
```

File: Sources/CovidCertificateSDK/ChCovidCert.swift [hunk 5]
```diff
@@ -154,14 +155,81 @@ public struct ChCovidCert {
...
+                          let rules = list.getRulesJSON(),
+                          let valueSets = list.getValueSetsJSON() else {
+                        completionHandler(.failure(.NETWORK_PARSE_ERROR))
+                        return
+                    }
+                    if case .failure = result.updateData(rules: rules, valueSets: valueSets) {
```
Comment (ubamrein) [inline Sources/CovidCertificateSDK/ChCovidCert.swift:174]:
```suggestion
                    if case .failure = certLogic.updateData(rules: rules, valueSets: valueSets) {
```

File: Sources/CovidCertificateSDK/NationalRules/NationalRulesVerifier.swift [hunk 6]
```diff
@@ -41,6 +42,7 @@ public enum NationalRulesError: Error, Equatable {
         case let .NETWORK_ERROR(code): return code.count > 0 ? "NE|\(code)" : "NE"
         case .NETWORK_PARSE_ERROR: return "NE|PE"
         case .NETWORK_NO_INTERNET_CONNECTION: return "NE|NIC"
+        case .UNKNOWN_TEST_FAILURE: return "NE|UKN"
```
Comment (ubamrein) [inline Sources/CovidCertificateSDK/NationalRules/NationalRulesVerifier.swift:45]: adjust NE to correct errorcode

File: Sources/CovidCertificateSDK/ehn/EuHealthCert.swift [hunk 7]
```diff
@@ -138,6 +138,23 @@ public struct Vaccination: Codable {
         }
     }
 
+    public func getValidFromDate(daysAfterFirstShot: Int) -> Date? {
```
Comment (ubamrein) [inline Sources/CovidCertificateSDK/ehn/EuHealthCert.swift:141]: unit test should check date

File: Sources/CovidCertificateSDK/ehn/EuHealthCert.swift [hunk 8]
```diff
@@ -147,6 +164,14 @@ public struct Vaccination: Codable {
         return date
     }
 
+    public func getValidUntilDate(maximumValidityInDays: Int) -> Date? {
```
Comment (ubamrein) [inline Sources/CovidCertificateSDK/ehn/EuHealthCert.swift:167]: unit test should check date

File: Sources/CovidCertificateSDK/ehn/EuHealthCert.swift [hunk 9]
```diff
@@ -213,6 +238,18 @@ public struct Test: Codable {
         }
     }
 
+    public func getValidUntilDate(pcrTestValidityInHours: Int, ratTestValidityInHours: Int) -> Date? {
```
Comment (ubamrein) [inline Sources/CovidCertificateSDK/ehn/EuHealthCert.swift:241]: unit test should check date

**Review by UBaggeler — APPROVED** [review]

**Pull request merged**